From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: sin-ack <sin-ack@protonmail.com>
Date: Tue, 27 Sep 2022 00:08:46 +0000
Subject: [PATCH] [libcxx] Add support for SerenityOS

This commit teaches libc++ about what features are available in our
LibC, namely:
* locale support, by providing __locale_dir/support/serenity.h
* The number of errno constants defined by us is given by the value of
  the `ELAST` macro.
* Multithreading is implemented though the pthread library.
* Use libc++'s builtin character type table instead of the one provided
  by LibC as there's a lot of extra porting work to convince the rest of
  locale.cpp to use our character type table properly.

This commit is an adaptation of the LLVM patch by Daniel Bertalan to fit
the layout of the zig-bootstrap project.

Co-Authored-By: Daniel Bertalan <dani@danielbertalan.dev>
Co-Authored-By: Linus Groh <mail@linusgroh.de>
---
 zig/lib/libcxx/include/__config               |  5 +++--
 .../include/__locale_dir/locale_base_api.h    |  2 ++
 .../__locale_dir/locale_base_api/serenity.h   | 22 +++++++++++++++++++
 .../libcxx/include/__locale_dir/messages.h    |  2 +-
 zig/lib/libcxx/src/include/config_elast.h     |  2 ++
 5 files changed, 30 insertions(+), 3 deletions(-)
 create mode 100644 zig/lib/libcxx/include/__locale_dir/locale_base_api/serenity.h

diff --git a/zig/lib/libcxx/include/__config b/zig/lib/libcxx/include/__config
index cc7a377334a83cbb34b8a997151b7e76b3810a9c..0544e206b3b1e66af15d71540a0f62545cd3b698 100644
--- a/zig/lib/libcxx/include/__config
+++ b/zig/lib/libcxx/include/__config
@@ -634,7 +634,8 @@ typedef __char32_t char32_t;
         defined(__APPLE__) ||                                                                                          \
         defined(__MVS__) ||                                                                                            \
         defined(_AIX) ||                                                                                               \
-        defined(__EMSCRIPTEN__)
+        defined(__EMSCRIPTEN__) ||                                                                                     \
+        defined(__serenity__)
 // clang-format on
 #      undef _LIBCPP_HAS_THREAD_API_PTHREAD
 #      define _LIBCPP_HAS_THREAD_API_PTHREAD 1
@@ -709,7 +710,7 @@ typedef __char32_t char32_t;
 #  endif
 
 #  if defined(__BIONIC__) || defined(__NuttX__) || defined(__Fuchsia__) || defined(__wasi__) ||                        \
-      _LIBCPP_HAS_MUSL_LIBC || defined(__OpenBSD__) || defined(__LLVM_LIBC__)
+      _LIBCPP_HAS_MUSL_LIBC || defined(__OpenBSD__) || defined(__LLVM_LIBC__) || defined(__serenity__)
 #    define _LIBCPP_PROVIDES_DEFAULT_RUNE_TABLE
 #  endif
 
diff --git a/zig/lib/libcxx/include/__locale_dir/locale_base_api.h b/zig/lib/libcxx/include/__locale_dir/locale_base_api.h
index 0474c1db359dea406b4313e22fe12daaf6c61aad..f9a48197ae60e75818dc000295135806d9e6734b 100644
--- a/zig/lib/libcxx/include/__locale_dir/locale_base_api.h
+++ b/zig/lib/libcxx/include/__locale_dir/locale_base_api.h
@@ -132,6 +132,8 @@
 #      include <__locale_dir/locale_base_api/ibm.h>
 #    elif defined(__OpenBSD__)
 #      include <__locale_dir/locale_base_api/openbsd.h>
+#    elif defined(__serenity__)
+#      include <__locale_dir/locale_base_api/serenity.h>
 #    endif
 
 #    include <__locale_dir/locale_base_api/bsd_locale_fallbacks.h>
diff --git a/zig/lib/libcxx/include/__locale_dir/locale_base_api/serenity.h b/zig/lib/libcxx/include/__locale_dir/locale_base_api/serenity.h
new file mode 100644
index 0000000000000000000000000000000000000000..b553ca2eb2110d0f78cb7706001316ab292569bf
--- /dev/null
+++ b/zig/lib/libcxx/include/__locale_dir/locale_base_api/serenity.h
@@ -0,0 +1,22 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _LIBCPP___LOCALE_LOCALE_BASE_API_SERENITY_H
+#define _LIBCPP___LOCALE_LOCALE_BASE_API_SERENITY_H
+
+#include <stddef.h>
+
+#include <locale.h>
+#include <__support/xlocale/__posix_l_fallback.h>
+#include <__support/xlocale/__strtonum_fallback.h>
+#include <clocale>
+#include <cstdlib>
+#include <ctype.h>
+#include <cwctype>
+
+#endif // _LIBCPP___LOCALE_LOCALE_BASE_API_SERENITY_H
diff --git a/zig/lib/libcxx/include/__locale_dir/messages.h b/zig/lib/libcxx/include/__locale_dir/messages.h
index 686f472840c222226b2e53fa1e3757c47ded8f13..b2c4a6fc3c60084e2450c7fb210d6eb021c05911 100644
--- a/zig/lib/libcxx/include/__locale_dir/messages.h
+++ b/zig/lib/libcxx/include/__locale_dir/messages.h
@@ -22,7 +22,7 @@
 
 #  if defined(__unix__) || (defined(__APPLE__) && defined(__MACH__))
 // Most unix variants have catopen.  These are the specific ones that don't.
-#    if !defined(__BIONIC__) && !_LIBCPP_LIBC_NEWLIB && !defined(__EMSCRIPTEN__)
+#    if !defined(__BIONIC__) && !_LIBCPP_LIBC_NEWLIB && !defined(__EMSCRIPTEN__) && !defined(__serenity__)
 #      define _LIBCPP_HAS_CATOPEN 1
 #      include <nl_types.h>
 #    else
diff --git a/zig/lib/libcxx/src/include/config_elast.h b/zig/lib/libcxx/src/include/config_elast.h
index be665a97bf91b5ade7d97b64f6a6b371514260b3..d0919e270d92e283aa6a1ba0a666c1e75d7f929a 100644
--- a/zig/lib/libcxx/src/include/config_elast.h
+++ b/zig/lib/libcxx/src/include/config_elast.h
@@ -37,6 +37,8 @@
 #  define _LIBCPP_ELAST 4095
 #elif defined(__APPLE__)
 // No _LIBCPP_ELAST needed on Apple
+#elif defined(__serenity__)
+// No _LIBCPP_ELAST needed on SerenityOS
 #elif defined(__MVS__)
 #  define _LIBCPP_ELAST 1160
 #elif defined(_LIBCPP_MSVCRT_LIKE)
